### ISSUE TITLE
Lint to disallow using `as` for pointer casts in favor of `.cast()`

### ISSUE THREAD

**Issue by retep998** [rust-lang/rust-clippy#5890]
### What it does

Any raw pointer cast from `*const T` to `*const U` or `*mut T` to `*mut U` that uses `as` should be caught by this lint. Instead you should use `.cast()`.

Should probably be a pedantic lint.

### Drawbacks

None.

### Example

```rust
foo as *const c_void
```

Could be written as:

```rust
foo.cast()
```

**Comment by matthiaskrgr** [comment]
There's a lint that seems to lint on `as` in general: https://rust-lang.github.io/rust-clippy/master/index.html#as_conversions

**Comment by retep998** [comment]
Yes, and there are also lints for specific types of as casts. There is no lint for specifically pointer casts while still allowing other kinds of as casts.

**Comment by ErichDonGubler** [comment]
This would seem to affect the [`transmutes_expressible_as_ptr_casts`](https://rust-lang.github.io/rust-clippy/master/#transmutes_expressible_as_ptr_casts) lint -- that link suggests using `as`, e.g., `p as *const [u16];`.

**Comment by rail-rain** [comment]
It seems to me this lint has already suggested twice (#4065 and https://github.com/rust-lang/rust-clippy/issues/4708#issuecomment-544371294). Maybe 4065 can be closed since 5890 has a bit more details.
